go-backend/db/user.go

```diff
@@ -2,21 +2,135 @@ package db
 }
 
+// Organization - retrieve the user's organization based on the OrgID property
+func (u *User) Organization(ctx context.Context, store Storer) (org Organization, err error) {
+	org, err = store.GetOrganization(ctx, u.OrgID)
+	return
+}
+
+// Role - retrieve the user's role based on the RoleID property
+func (u *User) Role(ctx context.Context, store Storer) (role Role, err error) {
+	role, err = store.GetRoleByID(ctx, u.RoleID)
+	return
+}
+
+// GetUserByEmail - Given an email address, return that user.
+func (s *pgStore) GetUserByEmail(ctx context.Context, email string) (user User, err error) {
+	err = s.db.Get(&user, `SELECT * FROM users WHERE email=$1 LIMIT 1`, email)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			err = ae.ErrRecordNotFound
+		}
+		// Possible that there's no rows in the result set
+		logger.WithField("err", err.Error()).Error("Error selecting user from database by email " + email)
+		return
+	}
+
+	// Populate the user's organization data
+	org := Organization{}
+	err = s.db.Get(&org, `SELECT * FROM organizations WHERE id=$1 LIMIT 1`, user.OrgID)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting user's organization from database by email and id " + email + " " + string(user.OrgID))
+	}
+	return
+}
+
+// GetUserByID - Given the database ID for that user, look them up.
+func (s *pgStore) GetUserByID(ctx context.Context, id int) (user User, err error) {
+	err = s.db.Get(&user, `SELECT * FROM users WHERE id=$1 LIMIT 1`, id)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting user from database by id " + string(id))
+		return
+	}
+	// Populate the user's organization data
+	org := Organization{}
+	err = s.db.Get(&org, `SELECT * FROM organizations WHERE id=$1`, user.OrgID)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting user's organization from database by user id and org id " + string(id) + " " + string(user.OrgID))
+	}
+	return
+}
+
+// ListUsers - retrieves all users from the database. Could be a very large result set...
 func (s *pgStore) ListUsers(ctx context.Context) (users []User, err error) {
 	err = s.db.Select(&users, "SELECT * FROM users ORDER BY name ASC")
 	if err != nil {
 		logger.WithField("err", err.Error()).Error("Error listing users")
+	}
+	return
+}
+
+// CreateNewUser - creates a new user in the database
+func (s *pgStore) CreateNewUser(ctx context.Context, u User) (newUser User, err error) {
+	// First, make sure they're not already in the database because if they are, we can just return that user
+	newUser, err = s.GetUserByEmail(ctx, u.Email)
+	if err == nil {
+		// If there's already a user here err will be nil, so newUser is populated. Just return, don't
+		// bother with all the rest of this function. No need to log an error that doesn't exist, either.
 		return
 	}
 
+	// If we made it this far, they don't appear to be in the database yet.
+	q := `INSERT INTO users (
```

Comment:
can we move this query to top and declare as a constant. Each time this function is called a new variable is declared

Reply:
e3f05d6a5d3df5da9a0bd4fc520d90efe0817a9c